### PR TITLE
[Serialization] Preserve source order in serialization

### PR DESCRIPTION
**Pull request by jrose-apple**
We previously shied away from this in order to not *accidentally* depend on it, but it becomes interesting again with textual interfaces, which can certainly be read by humans. The cross-file order is the order of input files, which is at least controllable by users.

### CODE REVIEW

File: include/swift/Serialization/ModuleFile.h [hunk 1]
```diff
@@ -379,7 +379,7 @@ class ModuleFile
 
   using DeclIDVector = SmallVector<serialization::DeclID, 4>;
 
-  DeclIDVector EagerDeserializationDecls;
```
Comment (jrose-apple) [inline include/swift/Serialization/ModuleFile.h]: This was unused.

File: lib/IDE/ModuleInterfacePrinting.cpp [hunk 2]
```diff
@@ -497,7 +497,7 @@ void swift::ide::printSubmoduleInterface(
   // If the group name is specified, we sort them according to their source order,
   // which is the order preserved by getTopLevelDecls.
   if (GroupNames.empty()) {
-    std::sort(SwiftDecls.begin(), SwiftDecls.end(),
+    std::stable_sort(SwiftDecls.begin(), SwiftDecls.end(),
```
Comment (jrose-apple) [inline lib/IDE/ModuleInterfacePrinting.cpp]: @nkcsgexi Arguably we could not sort at all now, but this kept the test churn down. What do you think is the right thing to do?

**Comment by jrose-apple** [comment]
@swift-ci Please test

File: lib/IDE/ModuleInterfacePrinting.cpp (hunk 2 above)
Comment (nkcsgexi) [reply lib/IDE/ModuleInterfacePrinting.cpp]: we've heard no complaint about this ordering. I think we should keep sorting them in this way.

**Review by nkcsgexi — APPROVED** [review]
LGTM, Thanks!

File: lib/IDE/ModuleInterfacePrinting.cpp (hunk 2 above)
Comment (jrose-apple) [reply lib/IDE/ModuleInterfacePrinting.cpp]: Okay. We should probably sort extensions too, then, but that can come later.

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Review by DougGregor — APPROVED** [review]
I like this direction

**Pull request merged**